### DISCUSSION TITLE
Two official ride raters!

### DISCUSSION THREAD

**Discussion by ocalhoun6** [OpenRCT2/OpenRCT2#19532] [Feature requests]
In a normal RCT park, there's one invisible peep who's the ride rater, and he visits one ride at a time to determine that ride's ratings. And normally this works very well without any potential for slowing down the player's PC. (Okay, not really, but that's kind of how the algorithm works. Only one ride's stats are calculated at a time.)

*But* this can run into problems when you have any ride that takes a very long time to complete, or any ride that gets stuck in an infinite loop (because of cheats/shoestringing/ride merging or because of a coaster valleying out or something). Then the 'ride rater' gets stuck on that ride and never moves on to update the ratings of any other ride. Causing issues like #19528 and #4695.

My proposed solution here is that there should be *two* official ride raters in every park. The standard one we all know and love, but also a second one: an _impatient_ one. The impatient one will keep track of how long he's been working on any particular ride, and if the ride is too long (say, more than 10 minutes), he'll get bored, quit tracking that ride, and move on to the next one instead. (It would be fine if he keeps track of that time only in a temporary variable -- saving and then reloading the park might make him forget how long he's been on the ride, restarting that 10-minute timer. But that should be no big deal.)

You still want to have the patient ride rater around, though, because sometimes people make very long rides, and they still want stats calculated for those. So you want to have that thankless guy out there monitoring Mr. Bone's Wild Ride to determine its ratings, no matter how long it takes. *But* you also add the impatient ride rater, so that the stats of other rides can still be calculated while the first guy is busy with Mr. Bones.

This could potentially double the load from ride ratings calculations, but *only* double it -- no more. Unless I'm greatly misunderstanding how expensive those calculations are, it shouldn't cause a noticeable slowdown on any remotely modern hardware. (It would also be a fantastic place to use multi-thread processing, taking advantage of multi-core processors.) And in exchange for that extra resource usage, you get to fix a persistent old bug, *plus* as long as the two raters choose rides to rate in different order, it should also mean that ride ratings will usually get calculated significantly quicker than before.

Not sure if this would be network safe, though? The behavior could be disabled during multiplayer if it poses a problem. It would still be very helpful in single player.

**Comment by Broxzier** [comment]
This idea sounds nice, but is technically more complex. I'd rather see the rating system being determined individually for each ride, instead of it being one task that iterates the rides one by one, and have it be refreshed occasionally. In the case of rides where there's an infinite loop, we might be able to detect these and abort the ratings for that ride altogether (though with plug-ins now able to modify the world, that may not be the best solution either).
